sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -464,6 +465,27 @@ object OptimizeIn extends Rule[LogicalPlan] {
 }
 
 /**
+  * Convert an expression into its conjunctive normal form (CNF), i.e. AND of ORs.
+  * For example, a && b || c is normalized to (a || c) && (b || c) by this method.
+  *
+  * Refer to https://en.wikipedia.org/wiki/Conjunctive_normal_form for more information
+  */
+object CNFNormalization extends Rule[LogicalPlan] {
+  def apply(plan: LogicalPlan): LogicalPlan = plan transform {
+    case q: LogicalPlan => q transformExpressionsUp {
+      case or @ Or(left, right) => (left, right) match {
```

Comment:
I would not make this a nested match as I think it makes it unnecessarily hard to read (and if you just use transform directly you won't have to manually handle the default case)